lib/mm.rb

```diff
@@ -0,0 +1,78 @@
+require_relative 'mm_code'
+require_relative 'mm_outputs'
+require_relative 'mm_input'
+require_relative 'mm_gen_list'
+
+module Mastermind
+    def self.play(blank_list = [], new_list = GeneratedList.new, new_code = Code.new(new_list.random_code), is_new = true)
+        previous_guesses = blank_list
+        remaining_guesses = 10
+        pins_out = blank_list
+        gen_list = new_list
+        game_code = new_code
+        game_out = Outputs.new
+        first_launch = is_new
+        until game_code.matched do
+            system "clear"
+            if first_launch
+                game_out.welcome
+                first_launch = false
+            end
+            if remaining_guesses.zero?
+                puts "You ran out of guesses..."
+                break
+            end
+            game_out.remaining_guesses(remaining_guesses)
+            if !previous_guesses.empty?
+                game_out.old_guesses(previous_guesses)
+            end
+            puts "Your color choices are R, G, Y, B, M, and C."
+            user_in = Input.new
+            until user_in.is_valid do
+                if user_in.restart_request == true
+                    restart("Are you sure you'd like to restart?", true)
+                elsif user_in.show_help
+                    game_out.available_cmds
+                elsif user_in.show_rules
+                    game_out.rules
+                elsif user_in.err_data[:err] == true
+                    system "clear"
+                    game_out.remaining_guesses(remaining_guesses)
+                    puts ""
+                    game_out.input_err(user_in.err_data)
+                    puts ""
+                end
+                user_in = Input.new
+            end
+            current_code = Code.new(user_in.user_input)
+            pins_out = game_code.compare(current_code.value)
+            remaining_guesses -= 1
+            previous_guesses << [(10 - remaining_guesses), current_code.convert_to_color, pins_out]
+        end
+        system "clear"
+        if game_code.matched
+            game_out.win
+            puts ("You won in only #{(10 - remaining_guesses).to_s} guesses!")
+        end
+        puts "The code was: " + game_code.convert.join(' ')
+        restart("Would you like to play again?", false)
+    end
+
+    def self.restart(string, cancel)
+        puts "#{string} (Y/n)"
+        confirm_restart = gets.chomp.upcase
+        if confirm_restart == "" || confirm_restart == "Y"
+            system "clear"
+            new_list = GeneratedList.new
+            new_code = Code.new(new_list.random_code)
```

Comment:
This namespace is initializing a lot of classes directly. That may imply that you're losing out on the benefits of [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection).

Ideally, if you're placing logic in other classes, that should imply that those dependencies can be replaced by some other class that implements the same set of methods but behaves differently. But it's hard to reap those benefits if you're initializing the dependency in the source code that's using it.

For example, you might have one `Outputs` class where the text is in English, and then another where the text is in Spanish. Your plan is to choose to use one or the other and swap it in on demand. In order to do so, you'd want to initialize that class outside of the namespace and then inject it based on which language was chosen for this game.